[compiler]add ccl bufferization and lccl dialect.

linuxlonelyeagle opened the pull request:
Add bufferization of ccl dialect and lcc dialect. see the test for the bufferization part.

qingyunqu commented on compiler/dialects/include/byteir/Dialect/Ccl/IR/CclOps.td:
```diff
@@ -217,8 +215,6 @@ def Ccl_SendOp : Ccl_Op<"send", [
 
 def Ccl_RecvOp : Ccl_Op<"recv", [
-    SameOperandsAndResultElementType,
-    SameOperandsAndResultShape,
```
Don't need to remove these when bufferize to lccl dialect?

linuxlonelyeagle replied:
The dynamic case is considered here.

qingyunqu replied:
If you remove this, you should add check for `ShapedType` in verify function.

qingyunqu commented on compiler/dialects/lib/Dialect/Ccl/IR/CclOps.cpp:
```diff
@@ -125,9 +125,6 @@ AllReduceOp::inferReturnTypes(MLIRContext *, std::optional<Location> location,
                               ValueRange operands, DictionaryAttr,
                               OpaqueProperties, RegionRange,
                               SmallVectorImpl<Type> &inferredReturnTypes) {
-  if (operands.size() != 1)
```
same to here.

qingyunqu commented on compiler/include/byteir/Dialect/Lccl/LcclOps.h:
```diff
@@ -0,0 +1,44 @@
+
+namespace mlir {
+namespace lccl {
+constexpr StringRef getRedOpSumName() { return "sum"; }
```
I think it's more simple to reuse ccl's definition.

qingyunqu commented on compiler/include/byteir/Dialect/Lccl/LcclOps.h:
```diff
@@ -0,0 +1,44 @@
+#include "mlir/IR/OpImplementation.h"
+
+using ReplicaGroupsIndices = llvm::SmallVector<int64_t>;
+using ReplicaGroupsIndicesRef = llvm::ArrayRef<int64_t>;
```
also reuse here.

linuxlonelyeagle commented on compiler/test/Dialect/Ccl/ccl_bufferize.mlir:
```diff
@@ -0,0 +1,215 @@
+// -----
+
+func.func @all_reduce_dynamic(%arg0: tensor<4xf32>, %arg1:tensor<1x4xi64>) -> tensor<4xf32> {
+    %0 = "ccl.all_reduce"(%arg0, %arg1) {reduction = "sum", synchronous=true}: (tensor<4xf32>, tensor<1x4xi64>) -> tensor<4xf32>
```
@qingyunqu Here is the case of all_reduce dynamic.

linuxlonelyeagle commented:
I added type inference for broadcast, all_gather,all_reduce,all_reduce_scatter. Added tighter restrictions in bufferize to further determine if there are indeterminate dimensions in the dynamic case, and if so, just stop running the pass.

qingyunqu reviewed (approved):
LGTM, thanks for the efforts.
